packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnTextCellData>> =
+        html<TableCellState<TableColumnTextCellData>>`
+            <nimble-text-field readonly="true" value=${x => x.data.value} placeholder=${x => x.placeholder}>
+            </nimble-text-field>
+        `;
+
+    public validateCellData(cellData: TCellData): boolean {
+        return typeof cellData['value'] === 'string';
+    }
+}
+```
+
+This also enables columns to access multiple pieces of data to use in its rendering:
+
+```TS
+type TableColumnNumberWithUnitCellData = NumberColumnData<'value'> & StringColumnData<'units'>;
+
+const format = (value: number, unit: string): string => {
+    return `${value.toString()} ${units}`;
+}
+
+public TableColumnNumberWithUnit extends FoundationElement implements ITableColumn<TableColumnNumberWithUnitCellData> {
+    ...
+
+    public cellDataKeyNames = ['value', 'units'] as const;
+
+    @attr
+    public valueKey: string;
+
+    @attr
+    public unitKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey, unitKey];
+    }
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnNumberWithUnitCellData>> =
+        html<TableCellState<TableColumnNumberWithUnitCellData>>`
+            <nimble-text-field
+                readonly="true"
+                value=${x => formatData(x.data.value, x.data.units)}
+            >
+            </nimble-text-field>
+        `;
+
+    public validateCellData(cellData: TCellData): boolean {
+        return typeof cellData['value'] === 'number' && typeof cellData['units'] === 'string';
+    }
+}
+```
+
+Here we have a column visualized in different ways based on custom logic:
+
+```TS
+type TableColumnPositiveNegativeNumberCellData = NumberColumnData<'value'>;
+
+const isPositive = (value: number): bool => {
+    return value >= 0;
+}
+
+public TableColumnPositiveNegativeNumber extends FoundationElement implements ITableColumn<TableColumnPositiveNegativeNumberCellData> {
+    ...
+
+    public cellDataKeyNames = ['value'] as const;
+
+    @attr
+    public valueKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public readonly cellStyles: ElementStyles = 
+        css`
+            .good {
+                color: green;
+            }
+
+            .bad {
+                color: red;
+            }
+        `;
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnPositiveNegativeNumberCellData>> =
+        html<TableCellState<TableColumnPositiveNegativeNumberCellData>`
+            html`
+                <nimble-text-field
+                    class=${_ => isPositive(x.data.value) ? "good" : "bad"}
+                    readonly="true"
+                    value=${x => x.data.value}
+                    style="color: red;"
```

Comment:
nit: don't need the inline style anymore

Reply:
Done.